### PR TITLE
Managed Iceberg dynamic destinations

### LINKED ISSUE #32365: [Task]: Add utilities to easily implement portable dynamic destinations
### What needs to happen?

Dynamic destinations is a feature that allows users to write to multiple destinations using one IO instance. This is usually done via some sort of function that inspects the record and determines its destination. Currently, this feature is only supported with native use-cases (e.g. Python can use [callables](https://beam.apache.org/releases/pydoc/current/apache_beam.io.fileio.html#dynamic-destinations); Java can use [SerializableFunction](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/FileIO.html#:~:text=as%20per%20above.-,Dynamic%20destinations,-If%20the%20elements)s or [DynamicDestinations](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.Write.html#to-org.apache.beam.sdk.io.gcp.bigquery.DynamicDestinations-)). 

Of course, we can’t simply pass a whole function from one language to another, but we can extract the important bits and articulate them in a language-agnostic way

After discussion in the [mailing list](https://lists.apache.org/thread/4oh3s94grpjypcfg0qzff4djqp8b1bff) and [design document](https://s.apache.org/portable-dynamic-destinations), the agreed upon approach (proposal (1) in the doc) looks something like this:

image

PR #31807 is a proof of concept that includes the necessary utilities and a real implementation for portable dynamic destinations in IcebergIO. This PR will be broken down into multiple PRs to decouple concepts and make it easier to review.

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by ahmedabu98**
Fixes #32365

This PR adds support for writing to dynamic Iceberg destinations, meaning that one transform instance can write to multiple Iceberg tables.

IcebergIO is offered as a Managed transform, which makes it a portable transform. Instead of offering a function of some sort, users can specify a String template and we will substitute in the appropriate record field values to determine the destination (see [RowStringInterpolator](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowStringInterpolator.java) for more details and examples)

For more information on portable dynamic destinations, see #32365

### CODE REVIEW

**Comment by ahmedabu98** [comment]
assign set of reviewers

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java [hunk 1]
```diff
@@ -160,7 +176,14 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
 
       IcebergIO.WriteRows writeTransform =
           IcebergIO.writeRows(configuration.getIcebergCatalog())
-              .to(TableIdentifier.parse(configuration.getTable()));
+              .to(
+                  new PortableIcebergDestinations(
```
Comment (chamikaramj) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:180]: Do we always expect the schema-transform to use "PortableIcebergDestinations" or will there be cases where we have to use the pass through (SingleTableDynamicDestinations) ?

**Review by chamikaramj — COMMENTED** [review]
Thanks!

LGTM

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java [hunk 2]
```diff
@@ -160,7 +176,14 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
...
+                      configuration.getTable(),
```
Comment (chamikaramj) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:181]: We should also update the documentation of table (string) to clarify how one can use a formatted string to specify a dynamic destination table.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/AssignDestinations.java [hunk 3]
```diff
@@ -41,23 +45,30 @@ public AssignDestinations(DynamicDestinations dynamicDestinations) {
   @Override
   public PCollection<Row> expand(PCollection<Row> input) {
 
-    final Schema inputSchema = input.getSchema();
     final Schema outputSchema =
         Schema.builder()
-            .addRowField("data", inputSchema)
-            .addRowField("dest", dynamicDestinations.getMetadataSchema())
+            .addStringField("dest")
```
Comment (chamikaramj) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/AssignDestinations.java:50]: These field names should go to consts somewhere (preferably in a proto).

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/PortableIcebergDestinations.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      String destinationTemplate,
+      String fileFormat,
+      Schema inputSchema,
+      @Nullable List<String> fieldsToDrop,
+      @Nullable List<String> fieldsToKeep,
+      @Nullable String onlyField) {
+    interpolator = new RowStringInterpolator(destinationTemplate, inputSchema);
+    RowFilter rf = new RowFilter(inputSchema);
+
+    if (fieldsToDrop != null) {
```
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/PortableIcebergDestinations.java:45]: The underlying RowFilter takes care of erroring out with helpful messages

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 1 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:180]: `PortableIcebergDestinations` also supports single table destination. Essentially if there are no curly braces in the destination string, it treats it like a single table.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 2 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:181]: We don't have any class documentation in `IcebergIO.java` yet. Can we add some in a future PR?

We can update the dataflow documentation for now though: https://cloud.google.com/dataflow/docs/guides/write-to-iceberg#configuration

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/AssignDestinations.java (hunk 3 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/AssignDestinations.java:50]: Sure will create constants. 

Do we need protos here? This code is just in Java

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 2 above)
Comment (chamikaramj) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:186]: Yeah, adding in a future PR is fine.

**Comment by chamikaramj** [comment]
Thanks. LGTM.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 2 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:186]: Adding documentation in #32621

**Pull request merged**